### ISSUE TITLE
Installing & deploying Tracim

### ISSUE THREAD

**Issue by grignards** [tracim/tracim#3088]
## Feature description

The goal of this issue is to collect the improvements we'd like to see to the installation and deployment of Tracim for three main cases:

- developing Tracim
- testing Tracim (CI tests)
- using Tracim

## Context

Installing and deploying Tracim is done by several means (manual installation with README, setup scripts, docker image, CI build scripts for tests) and it not a very simple task.

With the 3.0 version additional setup will be needed for TLMs (pushpin) and modifications of existing ones too (mail_notifier daemon). So we should take this chance to establish in which direction we want to go for installing Tracim.

## Ideas (brainstorm mode)

- make a pypi package of released versions (including frontend build ?)
- alternatively (for frontend) : make the released frontend builds available in github releases
- reduce the usage of bash for long scripts
- be able to generate the configuration file with a command (`tracimcli config init`)
- (docker image) use only one supervisor to start all components
- (docker image) use docker compose to split components (?)

**Comment by lebouquetin** [comment]
My point:

> make a pypi package of released versions (including frontend build ?)

This is a developer way of deploy a software, not a sysadmin way. I do not like this, by the way, because you can't patch easyly the source code (so if not possible to patch, the best way is to be "sysadmin-compliant")

> alternatively (for frontend) : make the released frontend builds available in github releases

I think we should have this. There is no value to build frontend on each deployment. For example, if you try to deploy tracim on a raspberry, you have 6 ours of installation - which is useless

> reduce the usage of bash for long scripts

Bash is the only standard tool available on a GNU/Linux system.

> be able to generate the configuration file with a command (tracimcli config init)

I do not see the interest for this. Could you explain? How will you manage migrations if you do not provide a configuration file without risking to overwrite current config? 

> (docker image) use only one supervisor to start all components
> (docker image) use docker compose to split components (?)

We use docker in a virtual machine way. This is not the philosophy of the tool *but* this is the way users use it. Splitting will make things more flexible but there is no need for this: if you need flexibility, you can deploy from sources.

Honestly, splitting docker image will be a bad evolution of ease-of-use.

> have a schema explaning the architecture of Tracim backend

This is an interesting thing.

Note: system administrators are part of our users. Do not decide a new way to deploy without having them in the loop. By the way, keep in mind that tracim deployers are not all sysadmin or developers but lots are simple "bidouilleurs".

**Comment by lebouquetin** [comment]
one reason for having a docker image is the `one commandline tracim test case`. We must keep this.

- If you thnk it will help, we can define use cases we want to address on how tracim can be deployed (on a functionnal way.
- Remember of the pets vs cattle allegory (we do not deploy thousands of tracim services - and our users do so)

**Comment by grignards** [comment]
> > reduce the usage of bash for long scripts
> 
> Bash is the only standard tool available on a GNU/Linux system.

Yes, but we could only "bootstrap" using bash then use python as soon as it is available (we need it for Tracim anyway).

> 
> > be able to generate the configuration file with a command (tracimcli config init)
> 
> I do not see the interest for this. Could you explain? How will you manage migrations if you do not provide a configuration file without risking to overwrite current config?

I was thinking of an help for configuring the first time Tracim is installed: the command asks some questions and generates the configuration file based on the user's answers.

**Comment by grignards** [comment]
> 
> > make a pypi package of released versions (including frontend build ?)
> 
> This is a developer way of deploy a software, not a sysadmin way. I do not like this, by the way, because you can't patch easyly the source code (so if not possible to patch, the best way is to be "sysadmin-compliant")
> 

I know that pypi is not the reference tool for sysadmins, but currently the tracim setup makes use of it (if not using the docker image). Having a pypi package should at least remove the need for installing development packages (we do install several *-dev debian packages which are irrelevant for installing a Tracim instance).

**Comment by lebouquetin** [comment]
>>> reduce the usage of bash for long scripts
>> Bash is the only standard tool available on a GNU/Linux system.
> Yes, but we could only "bootstrap" using bash then use python as soon as it is available (we need it for Tracim anyway).

This is a possible strategy. We should not rely on system-wide python as it may conflict with the python version we need.

>>> be able to generate the configuration file with a command (tracimcli config init)
>> I do not see the interest for this. Could you explain? How will you manage migrations if you do not provide a configuration file without risking to overwrite current config?
> I was thinking of an help for configuring the first time Tracim is installed: the command asks some questions and generates the configuration file based on the user's answers.

Ok. You mean something like a wizard? If soit may be a good idea, yes.

> I know that pypi is not the reference tool for sysadmins, but currently the tracim setup makes use of it (if not using the docker image). Having a pypi package should at least remove the need for installing development packages (we do install several *-dev debian packages which are irrelevant for installing a Tracim instance).

I'm afraid this forbid hot patching of production (even if hot-patching is not a nice way to do, it's sometime required and a great advantage to be able to do so)

----

Another important thing which is currently the case (and which was bad on your previous job @grignards ) : deployment must be based on same processus for both development environment and production.

This is a great advantage:

- allow to detect sccripts bug during development
- it avoid problems appearing on production
- it factorize engineering.

Currently, bash scripts used for development environment are the same as the ones used on docker image. This may appear not as efficient as possible for development environment but this is a good strategy for avoiding unexpected error case or side effects on production / execution.

things which are important to me as a target (we are not-so-far from this, I think):

- one command line dev env setup
- one command line test env run (test in the meaning of "a user wants to test")
- acceptable complexe deployment for complex cases (distributed, scalable deployement, specific cases, etc)

The docker image is designed for matching the second case. It is not intended to allow complex/scalable infrastructure - we would make something specific in this case)
Avoiding docker and other stuff for development seems to me a required limitation: source code being developed in python and javascript, python and javascript ecosystem stuff should be enough (eg: we do not need apache, we do not need redis, I think, except for workers, we even do not need postgresql).

The minimal dependencies we require, the easier a devenv can be setup, the more contributors we will have the more bugreports we will have too.

Last (but not least) : allowing disengageable mechanisms / dependencies is a good way to go. For exemple, when you auto-deploy with puppet, you are not anymore able to patch a given host. Patching is a bad habit but it's very usefull in some cases. Ansible allows to patch - you keekp control (and you can do wrong things - but you keep control).
